src/Workspaces/Core/Portable/Workspace/Solution/Checksum.cs

```diff
@@ -15,71 +16,125 @@ internal sealed partial class Checksum : IObjectWritable, IEquatable<Checksum>
     {
         public static readonly Checksum Null = new Checksum(Array.Empty<byte>());
 
-        private readonly byte[] _checkSum;
-        private int _lazyHash;
-
-        public Checksum(byte[] checksum)
+        [StructLayout(LayoutKind.Explicit, Size = 20)]
+        private struct Sha1Hash : IEquatable<Sha1Hash>
         {
-            // 0 means it is not initialized
-            _lazyHash = 0;
-            _checkSum = checksum;
-        }
+            [FieldOffset(0)]
+            public unsafe fixed byte Value[20];
 
-        public bool Equals(Checksum other)
-        {
-            if (other == null)
+            [FieldOffset(0)]
+            private long Data1;
+
+            [FieldOffset(8)]
+            private long Data2;
+
+            [FieldOffset(16)]
+            private int Data3;
+
+            public static bool operator ==(Sha1Hash x, Sha1Hash y)
             {
-                return false;
+                return x.Equals(y);
             }
 
-            if (_checkSum.Length != other._checkSum.Length)
+            public static bool operator !=(Sha1Hash x, Sha1Hash y)
             {
-                return false;
+                return !x.Equals(y);
+            }
+
+            public void WriteTo(ObjectWriter writer)
+            {
+                writer.WriteInt64(Data1);
+                writer.WriteInt64(Data2);
+                writer.WriteInt32(Data3);
+            }
+
+            public static Sha1Hash ReadFrom(ObjectReader reader)
+            {
+                Sha1Hash result = default;
+                result.Data1 = reader.ReadInt64();
+                result.Data2 = reader.ReadInt64();
+                result.Data3 = reader.ReadInt32();
+                return result;
             }
 
-            for (var i = 0; i < _checkSum.Length; i++)
+            public override int GetHashCode()
             {
-                if (_checkSum[i] != other._checkSum[i])
+                // The checksum is already a hash. Just read a 4-byte value to get a well-distributed hash code.
+                return (int)Data1;
+            }
+
+            public override bool Equals(object obj)
+            {
+                if (!(obj is Sha1Hash other))
                 {
                     return false;
                 }
+
+                return Equals(other);
```

Comment:
I'm torn on this. `Equals` methods are notoriously hard to read, so I rely on patterns to be consistent. But this expression seems straightforward enough that maybe it could be the new pattern.

On a separate note, I wonder which of the following is more efficient...

```csharp
if (obj is Checksum)
  return Equals((Checksum)obj);
else
  return false;
```

or...

```csharp
var other = obj as Sha1Hash?;
if (other != null)
  return Equals(other.Value);
else
  return false;
```